File: sdk/ios/Sources/FrakSDK/AppLink/ReferralArrival.swift
import Foundation

enum ReferralArrival {
    // Must be checked before tracking any arrival. Ignores a self-referral (this device's own
    // anonymousId as the link's clientId) and a foreign-merchant v2 context. A v1 link carries
    // neither field, so neither guard applies to it — that gap is known and still open.
    //
    // ownMerchantId is best-effort, from the cached config: nil means "unknown" and lets the
    // context through. Compared case-insensitively, or a casing difference drops a real referral.
    static func shouldIgnoreArrival(
        _ context: FrakContext,
        anonymousId: String?,
        ownMerchantId: String? = nil
    ) -> Bool {
        switch context {
        // No merchantId on a v1 context to compare against ownMerchantId — see the doc above.
        case .v1: false
        case .v2(let context):
            (anonymousId != nil && context.clientId == anonymousId)
                || (ownMerchantId.map { !Self.sameMerchant(context.merchantId, $0) } ?? false)
        }
    }

    static func sameMerchant(_ a: String, _ b: String) -> Bool {
        a.trimmingCharacters(in: .whitespaces).caseInsensitiveCompare(b.trimmingCharacters(in: .whitespaces))
            == .orderedSame
    }

    static func arrival(from context: FrakContext) -> Interaction {
        switch context {
        case .v1(let wallet):
            .arrival(referrerWallet: wallet)
        case .v2(let context):
            .arrival(
                referrerWallet: context.wallet,
                referrerClientId: context.clientId,
                referrerMerchantId: context.merchantId,
                referralTimestamp: context.timestamp
            )
        }
    }
}

File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/applink/ReferralArrival.kt
package id.frak.sdk.applink

import id.frak.sdk.sharing.FrakContext
import id.frak.sdk.tracking.Interaction

/** Turning an inbound referral context into an arrival, and deciding whether to. */
internal object ReferralArrival {
    /**
     * Mandatory before any arrival is tracked. Ignores a self-referral (this device's own
     * [anonymousId] as the link's `clientId`) and a foreign-merchant V2 referral; a V1 link carries
     * no `merchantId`, so it counts as this merchant's. A null [ownMerchantId] lets it through.
     */
    fun shouldIgnoreArrival(
        context: FrakContext,
        anonymousId: String?,
        ownMerchantId: String? = null,
    ): Boolean =
        when (context) {
            // A V1 context has no merchantId to compare against ownMerchantId.
            is FrakContext.V1 -> {
                false
            }

            is FrakContext.V2 -> {
                val sameDevice = anonymousId != null && context.clientId == anonymousId
                val foreignMerchant = ownMerchantId != null && !sameMerchant(context.merchantId, ownMerchantId)
                sameDevice || foreignMerchant
            }
        }

    /** Trim + case-insensitive: shared with [id.frak.sdk.tracking.InteractionSender]'s send-time guard. */
    internal fun sameMerchant(
        a: String,
        b: String,
    ): Boolean = a.trim().equals(b.trim(), ignoreCase = true)

    fun arrivalFrom(context: FrakContext): Interaction =
        when (context) {
            is FrakContext.V1 -> {
                Interaction.arrival(
                    referrerWallet = context.wallet,
                    // V1 carries an address and nothing else
                    referrerClientId = null,
                    referrerMerchantId = null,
                    referralTimestamp = null,
                )
            }

            is FrakContext.V2 -> {
                Interaction.arrival(
                    referrerWallet = context.wallet,
                    referrerClientId = context.clientId,
                    referrerMerchantId = context.merchantId,
                    referralTimestamp = context.timestamp,
                )
            }
        }
}
